### PR TITLE
Decouple catalog from RelAlgDispatcher with MetaData part II

### PR DESCRIPTION
**Pull request by JiyuInter**

### CODE REVIEW

File: QueryEngine/CiderCodeGenerator.cpp [hunk 1]
```diff
@@ -2860,9 +2860,9 @@ CiderCodeGenerator::compileWorkUnit(const std::vector<InputTableInfo>& query_inf
         (cgen_state_->filter_func_ ? serialize_llvm_object(cgen_state_->filter_func_)
                                    : "");
 
-#ifndef NDEBUG
-    llvm_ir += serialize_llvm_metadata_footnotes(query_func, cgen_state_.get());
-#endif
+// #ifndef NDEBUG
+//     llvm_ir += serialize_llvm_metadata_footnotes(query_func, cgen_state_.get());
```
Comment (winningsix) [inline QueryEngine/CiderCodeGenerator.cpp:2864]: uncomment this please and use cider::serialize_llvm_metadata_footnotes(query_func, cgen_state_.get());

File: QueryEngine/CiderCodeGenerator.h [hunk 2]
```diff
@@ -193,3 +193,4 @@ class CiderCodeGenerator {
   std::unordered_map<unsigned, AggregatedResult> leaf_results_;
   static const size_t code_cache_size{1000};
 };
+
```
Comment (winningsix) [inline QueryEngine/CiderCodeGenerator.h:196]: remove this.

File: QueryEngine/CiderRelAlgDispatcher.cpp [hunk 3]
```diff
@@ -414,6 +414,10 @@ std::vector<std::shared_ptr<RelAlgNode>> CiderRelAlgDispatcher::run(
       ra_node = dispatchLogicalValues(crt_node);
     } else if (rel_op == std::string("LogicalUnion")) {
       ra_node = dispatchUnion(crt_node);
+    } else if (rel_op == std::string("EnumerableTableScan") ||
+        rel_op == std::string("LogicalTableScan")) {
```
Comment (winningsix) [inline QueryEngine/CiderRelAlgDispatcher.cpp:418]: indent to 2 please

File: QueryEngine/CiderRelAlgDispatcher.cpp [hunk 4]
```diff
@@ -716,3 +733,25 @@ std::shared_ptr<const RelAlgNode> CiderRelAlgDispatcher::prev(
   CHECK_EQ(static_cast<size_t>(id), nodes_.size());
   return nodes_.back();
 }
+
+const TableDescriptor* getTableFromScanNode(const rapidjson::Value& scan_ra, MetaDesc meta) {
```
Comment (winningsix) [inline QueryEngine/CiderRelAlgDispatcher.cpp:737]: const for MetaDesc

File: QueryEngine/CiderRelAlgDispatcher.cpp [hunk 5]
```diff
@@ -716,3 +733,25 @@ std::shared_ptr<const RelAlgNode> CiderRelAlgDispatcher::prev(
...
+  const auto& table_json = field(scan_ra, "table");
+  CHECK(table_json.IsArray());
+  CHECK_EQ(unsigned(2), table_json.Size());
+  const auto td = meta.getData(table_json[1].GetString());
+  CHECK(td);
+  return td;
+}
+
+void check_empty_inputs_field(const rapidjson::Value& node) {
+  std::cout<<"3"<<std::endl;
```
Comment (winningsix) [inline QueryEngine/CiderRelAlgDispatcher.cpp:748]: remove this.

File: QueryEngine/CiderRelAlgDispatcher.cpp [hunk 6]
```diff
@@ -716,3 +733,25 @@ std::shared_ptr<const RelAlgNode> CiderRelAlgDispatcher::prev(
...
+  const auto& inputs_json = field(node, "inputs");
+  CHECK(inputs_json.IsArray() && !inputs_json.Size());
+}
+std::vector<std::string> getFieldNamesFromScanNode(const rapidjson::Value& scan_ra) {
+  std::cout<<"4"<<std::endl;
```
Comment (winningsix) [inline QueryEngine/CiderRelAlgDispatcher.cpp:753]: the same as above.

File: QueryEngine/CiderRelAlgDispatcher.cpp [hunk 7]
```diff
@@ -716,3 +733,25 @@ std::shared_ptr<const RelAlgNode> CiderRelAlgDispatcher::prev(
...
+  const auto& fields_json = field(scan_ra, "fieldNames");
+  std::cout<<"5"<<std::endl;
```
Comment (winningsix) [inline QueryEngine/CiderRelAlgDispatcher.cpp:755]: the same as above.

File: QueryEngine/CiderRelAlgDispatcher.h [hunk 8]
```diff
@@ -22,20 +22,23 @@
...
 #include "JsonAccessors.h"
 #include "RelAlgDagBuilder.h"
+#include "catalogApi.h"
 
 #include <memory>
 #include <string>
 #include <unordered_set>
 #include <vector>
 
+
```
Comment (winningsix) [inline QueryEngine/CiderRelAlgDispatcher.h:32]: remove this.

**Review by winningsix — COMMENTED** [review]
A few minor comments around code style.

File: QueryEngine/CiderRelAlgDispatcher.cpp (hunk 4 above)
Comment (jikunshang) [reply QueryEngine/CiderRelAlgDispatcher.cpp:737]: const reference.

File: QueryEngine/CiderRelAlgDispatcher.cpp [hunk 9]
```diff
@@ -392,7 +392,7 @@ JoinType to_join_type(const std::string& join_type_name) {
 }
 
 std::vector<std::shared_ptr<RelAlgNode>> CiderRelAlgDispatcher::run(
-    const rapidjson::Value& rels) {
+    const rapidjson::Value& rels, MetaDesc meta) {
```
Comment (winningsix) [inline QueryEngine/CiderRelAlgDispatcher.cpp:395]: const reference plz.

File: QueryEngine/CiderRelAlgDispatcher.cpp [hunk 10]
```diff
@@ -423,6 +426,19 @@ std::vector<std::shared_ptr<RelAlgNode>> CiderRelAlgDispatcher::run(
   return std::move(nodes_);
 }
 
+std::shared_ptr<RelScan> CiderRelAlgDispatcher::dispatchTableScan(const rapidjson::Value& scan_ra, MetaDesc meta) {
```
Comment (winningsix) [inline QueryEngine/CiderRelAlgDispatcher.cpp:429]: same as above.

File: QueryEngine/CiderRelAlgDispatcher.cpp [hunk 11]
```diff
@@ -716,3 +732,20 @@ std::shared_ptr<const RelAlgNode> CiderRelAlgDispatcher::prev(
   CHECK_EQ(static_cast<size_t>(id), nodes_.size());
   return nodes_.back();
 }
+
+const TableDescriptor* getTableFromScanNode(const rapidjson::Value& scan_ra, const MetaDesc meta) {
```
Comment (winningsix) [inline QueryEngine/CiderRelAlgDispatcher.cpp:736]: meta should be const reference.

**Review by winningsix — COMMENTED** [review]
Merged and please resolve comments in separated JIRA.

**Pull request merged**